### DISCUSSION TITLE
Task design upsampling

### DISCUSSION THREAD

**Discussion by Masharipov** [cvnlab/GLMsingle#132] [Q&A] [answered]
I wonder, is it possible to perform fractional rigid regression described in GLMsingle paper using fMRI time-series in original temporal resolution and task design, which was: 1) upsampled to TR/16, 2) convolved with voxel-specific HRF, 3) downsampled to TR  ? 
How do you estimate how much work will have to be done to create custom scripts for this goal based on GLMsingle functions (small changes or huge rework)? Maybe you can give me some tips on how to do this?

Im asking becouse "upsampling/convolution/downsampling" noticeably increases sensetivity and specificity of LSA and LSS methods for rapid event-related designs (I checked it on simulations). The same should be the case with GLMsingle approach. Upsampling of the fMRI time-series is not an option for two resons: it requeres huge computational resourses and it produces spurious results.

**Comment by kendrickkay** [comment]
I am not entirely sure what you are suggesting....  My guess is that you want to see if you can do some pre-computation of the regressors (involving upsampling, convolution, and then downsampling).   Also, I am not sure what you mean by 'spurious results'.

If your predictors already have the HRF convolution performed, you can call GLMsingle but specifying a HRF library that just consists of a delta function.  Convolving with a delta function will obviously have no effect, so it should essentially bypass the HRF library stuff.    Specifically, you can do:   wantlibrary => 0,  hrftoassume = 1.

**Comment by Masharipov** [comment]
Thanks for the quick response!

> I am not entirely sure what you are suggesting.... My guess is that you want to see if you can do some pre-computation of the regressors (involving upsampling, convolution, and then downsampling).

Ideally, I suggest to define stimulus onset times in seconds instead of TRs, create upsampled task regrssors, convolve them with upsampled voxel-specific HRF, then downsample convolved task regressor to fMRI resolution, as it is done in SPM, FSL, AFNI etc. There are many task designs with short ISI, jittering, short event duration (<TR) etc, which pose challenges in defining unconvolved task design using fMRI resolution.

>Also, I am not sure what you mean by 'spurious results'.

I was getting "sawtooth" runwiseFIR, but now I got rid of this problem. Now the problem is that calculating models with upsampled fMRI time series takes a very long time.

>If your predictors already have the HRF convolution performed, you can call GLMsingle but specifying a HRF library that just consists of a delta function. Convolving with a delta function will obviously have no effect, so it should essentially bypass the HRF library stuff. Specifically, you can do: wantlibrary => 0, hrftoassume = 1.

If I'm not interested in voxel-specific HRF, but only in how fractional rigid regression deals with multicollinearity problem compared to LSS and LSA, I guess I can bypass the HRF library.

Did I understand you correctly that I need:
1) Create convolved task design (e.g. in SPM) with separate regressors for each trial (as in LSA) and use it as _design_ variable.
2) Set _wantlibrary_ = 0
3) Set _hrftoassume_ = 1
4) Which _stimdur_ should I use? I guess it should be some dummy value (zero?), becouse I already calculated convolved task design

**Comment by kendrickkay** [comment]
Yeah, those steps are correct.  stimdur is unused in your case, so you can just set it to 1 or something.

**Comment by kendrickkay** [comment]
Whoops, on second thought, I just realized specifying pre-convolved stimulus predictors is not going to work with the current implementation of GLMsingle.  This is because GLMsingle has to handle the cross-validation stuff and currently that requires knowing about each trial onset, etc.   So, unfortunately if you want to proceed, you'll have to concoct a home-grown version of ridge regression as it applies to your current case.

Alternatively, upsampling is still a path forward.  If computational requirements are large, it appears you are doing things at TR/16 (where TR is 2 s or so?)?  In terms of the temporal properties of the HRF, there are diminishing benefits below a TR of, say, 1 s, or even 0.5 s.  Going finer-grain that that is probably not going to make any discernible difference.

**Reply by Masharipov** [reply]
> Whoops, on second thought, I just realized specifying pre-convolved stimulus predictors is not going to work with the current implementation of GLMsingle. This is because GLMsingle has to handle the cross-validation stuff and currently that requires knowing about each trial onset, etc. So, unfortunately if you want to proceed, you'll have to concoct a home-grown version of ridge regression as it applies to your current case.

I see, apparently it will be necessary to use the _fracridge_ function directly. Suppose I have convolved, prewhitened and filtered design matrix X and data Y. How can I implement cross-validation as in GLMsingle?

1) Split-half X and Y into {X1, X2} and {Y1, Y2}.
2) Calculate betas and alphas for the first half Y1:
fracs = 0.05 : 0.05 : 1;         
[coef1 alpha1]= fracridge(X1,fracs,Y1);
3) The next steps are not clear to me. I need to find the best alpha1 and use it for Y2 to calculate beta. How can I do this based on the results of _fracridge_?
4) Repeat the same, but swap Y1 and Y2
[coef2 alpha2]= fracridge(X2,fracs,Y2);
Find best alpha2, use it to calculate betas for Y1.

> Alternatively, upsampling is still a path forward. If computational requirements are large, it appears you are doing things at TR/16 (where TR is 2 s or so?)? In terms of the temporal properties of the HRF, there are diminishing benefits below a TR of, say, 1 s, or even 0.5 s. Going finer-grain that that is probably not going to make any discernible difference.

Actually, I found that upsampling (TR/16, TR = 2s) increased the sensetivity and specificity of LSA and LSS. This is evident for the sluggish canonical HRF (from SPM). If we reduce time-to-peak and FWHM of the HRF, the upsampling procedure becomes event more relevant. In the HRF library, time-to-peak can be reduced up to 4s. Also, the difference between individual HRFs from the library is pretty small (e.g. compare time-to-peak for HRF1 and HRF2), so the benefit of using HRF library may be somewhat diminished if upsampling is not used. I mean, we want to find difference in time-to-peak of about 0.1s between different HRFs and at the same time we perfom convolution in low resolution which blurs this tiny differnces.

**Comment by kendrickkay** [comment, marked as answer]
> I see, apparently it will be necessary to use the fracridge function directly. Suppose I have convolved, prewhitened and filtered design matrix X and data Y. How can I implement cross-validation as in GLMsingle?
> 
> Split-half X and Y into {X1, X2} and {Y1, Y2}.
> Calculate betas and alphas for the first half Y1:
> fracs = 0.05 : 0.05 : 1;
> [coef1 alpha1]= fracridge(X1,fracs,Y1);
> The next steps are not clear to me. I need to find the best alpha1 and use it for Y2 to calculate beta. How can I do this based on the results of fracridge?
> Repeat the same, but swap Y1 and Y2
> [coef2 alpha2]= fracridge(X2,fracs,Y2);
> Find best alpha2, use it to calculate betas for Y1.

Perhaps it would be best to review the existing GLMsingle code implementation (in either matlab or python) to see the innards of the algorithm.  In GLMsingle, cross-validation is done with respect to individual trials' betas.

**Reply by Masharipov** [reply]
Thanks for the help!
